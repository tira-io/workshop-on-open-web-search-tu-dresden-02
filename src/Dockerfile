# docker build -t bla src
#see https://github.com/OpenWebSearch/wows-code/tree/main/ecir24/baselines/dev-container
FROM webis/wows-2024-baseline:0.0.5

ENV PYTERRIER_VERSION='5.8'
ENV PYTERRIER_HELPER_VERSION='0.0.8'

RUN apt-get update \
	&& DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt-get install -y git openjdk-11-jdk curl wget gcc \
	&& pip3 install python-terrier pandas jupyterlab runnb tira \
	&& python3 -c "import pyterrier as pt; pt.init(version='${PYTERRIER_VERSION}', helper_version='${PYTERRIER_HELPER_VERSION}');" \
	&& python3 -c "import pyterrier as pt; pt.init(version='${PYTERRIER_VERSION}', helper_version='${PYTERRIER_HELPER_VERSION}', boot_packages=['com.github.terrierteam:terrier-prf:-SNAPSHOT']);"

# install trec-cast-tools for chunking of documents
RUN cd /home && git clone https://github.com/grill-lab/trec-cast-tools.git
RUN pip install -r /home/trec-cast-tools/corpus_processing/requirements.txt
ENV PYTHONPATH "${PYTHONPATH}:/home/trec-cast-tools/corpus_processing/"

# install colbert
RUN pip install --upgrade git+https://github.com/terrierteam/pyterrier_colbert.git

# install pytest
RUN pip install pytest approvaltests

#install sentence transformers
RUN pip install sentence_transformers

ADD snippet_generation.py /code/snippet_generation.py
ADD parameterized_spacy_passage_chunker.py /code/parameterized_spacy_passage_chunker.py

ENTRYPOINT [ "python3", "/code/snippet_generation.py" ]
