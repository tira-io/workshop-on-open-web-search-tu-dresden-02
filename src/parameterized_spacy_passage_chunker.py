# adapted from passage_chunkers.spacy_passage_chunker
# https://github.com/grill-lab/trec-cast-tools/blob/master/corpus_processing/passage_chunkers/spacy_passage_chunker.py

import spacy
from tqdm import tqdm

from passage_chunkers.abstract_passage_chunker import AbstractPassageChunker

nlp = spacy.load("en_core_web_sm", exclude=[
    "parser", "tagger", "ner", "attribute_ruler", "lemmatizer", "tok2vec"])
nlp.enable_pipe("senter")
nlp.max_length = 2000000  # for documents that are longer than the spacy character limit


class ParameterizedSpacyPassageChunker(AbstractPassageChunker):
    """
    Basically the same as #SpacyPassageChunker. Only difference is that the snippet size can be set in #__init__
    """
    def __init__(self, snippet_size=250):
        self.snippet_size = snippet_size

    def process_batch(self, document_batch: list[dict]) -> list[dict]:
        """
        Divides the documents of the given document list into snippets.
        :param document_batch: List of documents. Documents are dicts - content must be at 'contents' key.
        :return: List of documents. Documents are dicts - snippets can be found at 'contents' key.
        """
        batch_document_texts = [document['contents'] for document in document_batch]
        processed_document_texts = nlp.pipe(batch_document_texts, n_process=1)

        for index, document in tqdm(enumerate(processed_document_texts), total=len(document_batch)):
            document_sentences = list(document.sents)
            sentences_word_count = [
                len([token for token in sentence])
                for sentence in document_sentences
            ]

            generated_passages = self.chunk_document(document_sentences, sentences_word_count, self.snippet_size)
            document_batch[index]['contents'] = generated_passages

        return document_batch
