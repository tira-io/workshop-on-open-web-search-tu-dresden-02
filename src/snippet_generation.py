#!/usr/bin/env python3
import os
import argparse
import re
import shutil
from pathlib import Path

from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
import pandas as pd
import pyterrier as pt
import pyterrier_dr

from sentence_transformers import CrossEncoder
from src.parameterized_spacy_passage_chunker import ParameterizedSpacyPassageChunker
# Load a patched ir_datasets that loads the injected data inside the TIRA sandbox
from tira.third_party_integrations import load_rerank_data, ensure_pyterrier_is_loaded, get_output_directory

ensure_pyterrier_is_loaded()

tokeniser = pt.autoclass("org.terrier.indexing.tokenisation.Tokeniser").getTokeniser()
def pt_tokenise(text):
    return ' '.join(tokeniser.getTokens(text))

def split_dataframe_into_snippets(documents: pd.DataFrame, snippet_size=250) -> pd.DataFrame:
    document_list = documents.rename(columns={'text': 'contents'}).to_dict('records')

    chunker = ParameterizedSpacyPassageChunker(snippet_size)
    document_list = chunker.process_batch(document_list)

    return pd.DataFrame(document_list).rename(columns={'contents': 'text'})


def split_into_snippets(document_text: str, snippet_size=250) -> list[dict]:
    chunker = ParameterizedSpacyPassageChunker(snippet_size)
    return chunker.process_batch([{
        "id": 0,
        "url": '',
        "title": '',
        "contents": document_text
    }])[0]['contents']


def transform_snippet_format(snippets):
    df = pd.DataFrame({
        'docno': [str(snippet['id']) for snippet in snippets],
        'text': [snippet['body'] for snippet in snippets]
    })
    return df


def rank_snippets_lexical(query, snippets_df, ranker):
    pd_indexer = pt.DFIndexer(index_path="memory_index",type=pt.index.IndexingType.MEMORY)
    indexref3 = pd_indexer.index(snippets_df["text"], snippets_df["docno"])
    index = pt.IndexFactory.of(indexref3)
    retrieved = pt.BatchRetrieve(index, controls={"wmodel": ranker})

    #remove ? due to error in terrier query parser
    query = pt_tokenise(query)
    result = retrieved.search(query)

    merged_df = pd.merge(snippets_df, result, on='docno')
    merged_df = merged_df.sort_values('score', ascending=False)

    # Convert to list of dictionaries
    result_list = merged_df.apply(lambda row: {'score': row['score'], 'text': row['text']}, axis=1)

    return result_list.tolist()

def crossencode(query, top_k_snippets):
    top_k_texts = [d['text'] for d in top_k_snippets]
    #model = CrossEncoder('cross-encoder/ms-marco-MiniLM-L-6-v2', max_length=512)
    pairs = [(query, doc) for doc in top_k_texts]

    model = AutoModelForSequenceClassification.from_pretrained('cross-encoder/ms-marco-MiniLM-L-6-v2')
    tokenizer = AutoTokenizer.from_pretrained('cross-encoder/ms-marco-MiniLM-L-6-v2')

    features = tokenizer(pairs,  padding=True, truncation=True, return_tensors="pt")

    model.eval()
    with torch.no_grad():
        scores = model(**features).logits
        scores = scores.flatten().tolist()

    #scores = model.predict(pairs)
    reranked_top_k = [{'score': scores[i], 'text': top_k_texts[i]} for i in range(len(top_k_texts))]
    return reranked_top_k


def colbert_pipeline(docs_df: pd.DataFrame, query):
    colbert_model = pyterrier_dr.TctColBert('sentence-transformers/all-MiniLM-L12-v2')
    docs_df['qid'] = '0'
    docs_df['query'] = query
    #print(docs_df)
    result_df = colbert_model(docs_df)
    #print(result_df)
    merged_df = pd.merge(docs_df, result_df, on='docno')
    merged_df = merged_df.sort_values('score', ascending=False)
    #print(merged_df)

    # Convert to list of dictionaries
    result_list = merged_df.apply(lambda row: {'score': row['score'], 'text': row['text_x']}, axis=1)

    return result_list.tolist()


def find_top_snippets(query, snippets, ranker='Tf', max_snippets=3, snippet_size=250, use_crossencoder=True):
    # Check if document or query is empty
    regexp = re.compile(r'[a-zA-Z0-9]') 
    if not regexp.search(snippets) or not regexp.search(query):
        return []

    # First: split document_text into snippets
    # https://github.com/grill-lab/trec-cast-tools/tree/master/corpus_processing/passage_chunkers

    # Second: transform snippet format from output of split_into_snippets to input of rank_snippets

    snippets_df = transform_snippet_format(snippets)

    # Third: rank snippets

    if ranker in ('BM25', 'PL2', 'Tf'):
        ranking = rank_snippets_lexical(query, snippets_df, ranker)
        if use_crossencoder:
            ranking = crossencode(query, ranking[0:max_snippets])
    elif ranker == 'ColBERT':
        #non functional
        ranking = colbert_pipeline(snippets_df, [query])
        if use_crossencoder:
            ranking = crossencode(query, ranking[0:max_snippets])

    # Return values
    print(ranking)
    return ranking[0:max_snippets]


def parse_arguments():
    parser = argparse.ArgumentParser(description="Argument Parser")

    parser.add_argument("--retrieval", choices=["BM25", "PL2", "Tf", "ColBERT"], default="Tf", help="The retrieval "
                                                                                                    "model")
    parser.add_argument("--cross-encode", action="store_true", default=False, help="Use a cross-encoder to re-rank "
                                                                                   "the top-k passages of the "
                                                                                   "retrieval model")
    parser.add_argument("--snippet-size", default=250, type=int, help="The approximate size of created snippets")
    parser.add_argument("--top-snippets", default=3, type=int, help="Number k for top k snippets that are retrieved.")

    return parser.parse_args()


if __name__ == '__main__':
    # In the TIRA sandbox, this is the injected re-ranking dataset, injected via the environment variable TIRA_INPUT_DIRECTORY
    re_rank_dataset = load_rerank_data(default='workshop-on-open-web-search/re-ranking-20231027-training')
    print(re_rank_dataset)
    # Alternatively, you could use the scored docs of ir_datasets, e.g.:
    # from tira.third_party_integrations import ir_dataset
    # re_rank_dataset = ir_datasets.load(default='workshop-on-open-web-search/document-processing-20231027-training')

    args = parse_arguments()
    preprocessed_docs = split_into_snippets(re_rank_dataset)
    document_snippets = []
    for _, i in preprocessed_docs:
        document_snippets += [
            {'qid': i['qid'], 'docno': i['docno'], 'snippets': find_top_snippets(i['query'], i['text'], args.retrieval,
                                                                                 args.top_snippets, args.snippet_size,
                                                                                 args.cross_encode)}]

    document_snippets = pd.DataFrame(document_snippets)

    # The expected output directory, injected via the environment variable TIRA_OUTPUT_DIRECTORY
    output_dir = get_output_directory('.')

    output_file = Path(output_dir) / 'documents.jsonl.gz'
    document_snippets.to_json(output_file, lines=True, orient='records')