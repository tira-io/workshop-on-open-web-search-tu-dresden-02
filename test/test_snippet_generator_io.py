import unittest
from approvaltests import verify_as_json
from src.snippet_generation import split_into_snippets


class SnippetGeneratorIOTest(unittest.TestCase):
    test_documents = [
        "The Hubble telescope discovered two moons of Pluto, Nix and Hydra.",
        "Edwin Hubble, an astronomer with great achievement, completely reimagined our place in the universe (the "
        "telescope is named by him).",
        "Press ESC key, then the : (colon), and type the wq command after the colon and hit the Enter key to save and "
        "leave Vim.",
        "In Vim, you can always press the ESC key on your keyboard to enter the normal mode in the Vim editor.",
        "Common heart attack symptoms include: (1) Chest pain, (2) Pain or discomfort that spreads to the shoulder, "
        "arm, back, neck, jaw, teeth or sometimes the upper belly, etc.",
        "A heart attack happens when the flow of blood that brings oxygen to your heart muscle suddenly becomes "
        "blocked."
    ]

    def test_split_into_snippets_output(self):
        actual = [split_into_snippets(document) for document in self.test_documents]
        verify_as_json(actual)


if __name__ == '__main__':
    pass
    #unittest.main()
