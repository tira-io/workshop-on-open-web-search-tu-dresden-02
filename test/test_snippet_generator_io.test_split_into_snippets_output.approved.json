[
    {
        "body": "The Hubble telescope discovered two moons of Pluto, Nix and Hydra.",
        "id": 0
    },
    {
        "body": "Edwin Hubble, an astronomer with great achievement, completely reimagined our place in the universe (the telescope is named by him).",
        "id": 0
    },
    {
        "body": "Press ESC key, then the : (colon), and type the wq command after the colon and hit the Enter key to save and leave Vim.",
        "id": 0
    },
    {
        "body": "In Vim, you can always press the ESC key on your keyboard to enter the normal mode in the Vim editor.",
        "id": 0
    },
    {
        "body": "Common heart attack symptoms include: (1) Chest pain, (2) Pain or discomfort that spreads to the shoulder, arm, back, neck, jaw, teeth or sometimes the upper belly, etc.",
        "id": 0
    },
    {
        "body": "A heart attack happens when the flow of blood that brings oxygen to your heart muscle suddenly becomes blocked.",
        "id": 0
    }
]
